services/python_engine/header/rocketjoe/services/python_engine/celery.hpp

```diff
@@ -1,43 +1,400 @@
 #pragma once
 
-#include <pybind11/stl.h>
+#include <memory>
+#include <optional>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include <boost/functional/hash.hpp>
+#include <boost/heap/priority_queue.hpp>
+#include <boost/uuid/random_generator.hpp>
+#include <boost/uuid/uuid.hpp>
+
 #include <pybind11/functional.h>
 #include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+namespace rocketjoe::services::python_engine {
+
+    using namespace pybind11::literals;
+    namespace py = pybind11;
+
+    class celery;
+
+    class result_base {};
+
+    class eager_result : public result_base {
+    public:
+        template<typename T1, typename T2>
+        inline eager_result(T1 &&id, T2 &&ret_val);
+
+        inline auto get() -> py::object const;
+
+    private:
+        boost::uuids::uuid id;
+        py::object ret_val;
+    };
+
+    class async_result : public result_base {
+    public:
+        template<typename T1, typename T2, typename T3>
+        inline async_result(T1 &&id, T2 &&task_name, T3 &&app);
+
+        inline ~async_result();
+
+        inline auto get() -> py::object;
+
+        template<typename T1>
+        inline auto resolve(T1 &&ret_val) -> void;
+
+    private:
+        boost::uuids::uuid id;
+        std::string task_name;
+        py::object ret_val;
+        std::weak_ptr<celery> app;
+    };
+
+    class task {
+    public:
+        template<typename T1, typename T2, typename T3>
+        inline task(T1 &&task_name, T2 &&task_handler, T3 &&app);
```

Comment:
Why is there a &&  ? `T1 &&task_name, T2 &&task_handler, T3 &&app`

Reply:
I thought I'd try using the perfect forwarding pattern.